pkg/genericcli/context.go

```diff
@@ -0,0 +1,840 @@
+
+	return nil
+}
+
+func (c *ContextManager) setProject(args []string) error {
+	project, err := GetExactlyOneArg(args)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.Update(&ContextUpdateRequest{DefaultProject: &project})
+	if err != nil {
+		return err
+	}
+
+	_, _ = fmt.Fprintf(c.cfg.Out, msgProjectSwitched, successCheck(), color.GreenString(project))
+
+	return nil
+}
+
+func (c *ContextManager) ContextListCompletion(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	ctxs, err := c.getContexts()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+	var names []string
+	for _, ctx := range ctxs.Contexts {
+		names = append(names, ctx.Name)
+	}
+	return names, cobra.ShellCompDirectiveNoFileComp
+}
+
+func (c *ContextManager) writeContexts(ctxs *contexts) error {
+	if err := ctxs.validate(); err != nil {
+		return err
+	}
+	raw, err := yaml.Marshal(ctxs)
+	if err != nil {
+		return err
+	}
+
+	dest, err := c.cfg.configPath()
+	if err != nil {
+		return err
+	}
+
+	// when path is in the default path, we ensure the directory exists
+	defaultPath, err := c.cfg.defaultConfigDirectory()
+	if err != nil {
+		return fmt.Errorf(errMsgCannotGetDefaultDir, err)
+	}
+	if defaultPath == path.Dir(dest) {
+		err = c.cfg.Fs.MkdirAll(defaultPath, 0700)
+		if err != nil {
+			return fmt.Errorf(errMsgCannotEnsureDefaultDir, err)
+		}
+	}
+
+	err = afero.WriteFile(c.cfg.Fs, dest, raw, 0600)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *ContextConfig) configPath() (string, error) {
+	if viper.IsSet(keyConfig) {
+		return viper.GetString(keyConfig), nil
+	}
+
+	dir, err := c.defaultConfigDirectory()
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(dir, c.ConfigName), nil
+}
+
+func (c *ContextConfig) defaultConfigDirectory() (string, error) {
+	// TODO implement XDG specification?
+	h, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(h, "."+c.ConfigDirName), nil
+}
+
+func (c *ContextManager) getContexts() (*contexts, error) {
+	configPath, err := c.cfg.configPath()
+	if err != nil {
+		return nil, fmt.Errorf(errMsgCannotGetConfigPath, err)
+	}
+
+	raw, err := afero.ReadFile(c.cfg.Fs, configPath)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return &contexts{}, nil
+		}
+
+		return nil, fmt.Errorf(errMsgCannotReadConfig, c.cfg.ConfigName, err)
+	}
+
+	var ctxs contexts
+	err = yaml.Unmarshal(raw, &ctxs)
+
+	if ctxCurrent, ok := ctxs.getByName(ctxs.CurrentContext); ok {
+		ctxCurrent.IsCurrent = true
+	}
+
+	return &ctxs, err
+}
+
+func (c *ContextManager) Get(name string) (*Context, error) {
+	ctxs, err := c.getContexts()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, ok := ctxs.getByName(name)
+	if !ok {
+		return nil, fmt.Errorf(errMsgContextNotFound, name)
+	}
+	return ctx, nil
+}
+
+func (c *ContextManager) List() ([]*Context, error) {
+	ctxs, err := c.getContexts()
```

Comment:
Remove `getContexts` function and embed the implementation here. Try to rely on this method instead.